vng_api_common/caching/etags.py

```diff
@@ -1,79 +1,62 @@
 from django.http import HttpRequest
-from django.utils.translation import ugettext_lazy as _
 
 from djangorestframework_camel_case.render import CamelCaseJSONRenderer
-from rest_framework import mixins, serializers
-from rest_framework.exceptions import APIException
-from rest_framework.generics import GenericAPIView
+from rest_framework import serializers
 from rest_framework.request import Request
 from rest_framework.settings import api_settings
 
-from .utils import get_subclasses
+from ..serializers import GegevensGroepSerializer
+from ..utils import get_resource_for_path, get_subclasses
 
-CACHE_HEADER = "ETag"
 
+class StaticRequest(HttpRequest):
+    def get_host(self) -> str:
+        site = Site.objects.get_current()
+        return site.domain
 
-def has_cache_header(view: GenericAPIView) -> bool:
-    if view.request.method not in ("GET", "HEAD"):
-        return False
-
-    if hasattr(view, "detail") and not view.detail:
-        return False
-
-    if not isinstance(view, mixins.RetrieveModelMixin):
-        return False
-
-    return True
+    def _get_scheme(self) -> str:
+        return "https" if settings.IS_HTTPS else "http"
 
 
-class ETagMixin(models.Model):
-    """
-    Automatically calculate the (new) ETag value on save.
+def calculate_etag(instance: models.Model) -> str:
     """
+    Calculate the MD5 hash of a resource representation in the API.
 
-    _etag = models.CharField(
-        _("etag value"),
-        max_length=32,
-        help_text=_("MD5 hash of the resource representation in its current version."),
-        editable=False,
-    )
-
-    class Meta:
-        abstract = True
-
-    def save(self, *args, **kwargs):
-        self._etag = calculate_etag(self)
-        super().save(*args, **kwargs)
+    The serializer for the model class is retrieved, and then used to construct
+    the representation of the instance. Then, the representation is rendered
+    to camelCase JSON, after which the MD5 hash is calculated of this result.
+    """
+    model_class = type(instance)
+    serializer_class = _get_serializer_for_models()[model_class]
 
+    # build a dummy request with the configured domain, since we're doing STRONG
+    # comparison. Required as context for hyperlinked serializers
+    request = Request(StaticRequest())
+    request.version = api_settings.DEFAULT_VERSION
+    request.versioning_scheme = api_settings.DEFAULT_VERSIONING_CLASS()
 
-class APICache:
-    def __init__(self, view: GenericAPIView, etag_field: str = "_etag"):
-        """
-        API Cache interface.
+    serializer = serializer_class(instance=instance, context={"request": request})
 
-        :param view: the view or viewset instance currently used in the request
-        """
-        self.view = view
-        self.etag_field = etag_field
+    # render the output to json, which is used as hash input
+    renderer = CamelCaseJSONRenderer()
+    rendered = renderer.render(serializer.data, "application/json")
```

Comment:
Is dit niet een flinke performance hit? Waarom niet op model niveau?

Reply:
Omdat renderen als JSON vs. renderen als XML (of andere content-types) een andere ETag waarde moeten teruggeven conform best practices. Performance hit... verwacht dat dat wel meevalt. Hier en daar zijn er wat kleine tweaks te doen in `get_resource_for_path` door het in een `@lru_cache` te stoppen ook denk ik, maar ik profile het liever voor ik aannames doe.

Sowieso wordt deze code pas bij het saven uitgevoerd, en in alternatieve implementaties zouden we het kunnen off-loaden naar celery bijvoorbeeld.

Reply:
Check.